Refine Apache IoTDB bridge doc (#2089)
emqtt/emq-docs-cn: en_US/data-integration/data-bridge-iotdb.md

This page provides a comprehensive introduction to the data integration between EMQX and Apache IoTDB with practical instructions on creating and validating the data integration.

## How It Works

<!-- In this section, you explain how the data integration works together with the rule engine by introducing the architecture, emphasizing the simplicity and no coding required….You can also combine the working principle with a real-life use case. Some typical scenarios include connected vehicles, IIoT, power and energy, etc. You can describe where the data will be used eventually and what value they can bring to the business under the specific scenario. Use sequential steps to describe how data flows from devices to the data integration, and then to the data storage…-->

> The Apache IoTDB data integration is an out-of-the-box feature in EMQX designed to bridge the gap between raw MQTT-based time series data and IoTDB's powerful data storage capabilities. With a built-in [rule engine](https://docs.emqx.com/en/enterprise/v5.1/data-integration/rules.html) component, the integration simplifies the process of ingesting data from EMQX to IoTDB for storage and query, eliminating the need for complex coding.

Reviewer: As a suggestion, by referencing the relative path to the rule engine docs, we might avoid having links to older versions when we release a new version such as 5.2.

```suggestion
The Apache IoTDB data integration is an out-of-the-box feature in EMQX designed to bridge the gap between raw MQTT-based time series data and IoTDB's powerful data storage capabilities. With a built-in [rule engine](./rules.md) component, the integration simplifies the process of ingesting data from EMQX to IoTDB for storage and query, eliminating the need for complex coding.
```

The diagram below illustrates a typical architecture of data integration between EMQX and IoTDB. <!-- This image needs to be modified to be IoTDB specific-->

<img src="./assets/IoTDB_bridge_architecture.png" alt="IoTDB_bridge_architecture" style="zoom:67%;" />

The workflow of the data integration is as follows:

1. **Message publication from IoT devices**: Devices, whether they are part of connected vehicles, IIoT systems, or energy management platforms, send messages via MQTT based on their operational states, readings, or triggered events.
2. **Message reception by EMQX:** As an MQTT Broker, EMQX receives these MQTT messages from the IoT devices. It serves as the centralized hub for handling MQTT-based communication.
3. **Message data processing:** When a message arrives, it passes through the rule engine and is then processed by the rule defined in EMQX. The rules, based on predefined criteria, determine which messages need to be routed to IoTDB. If any rules specify payload transformations, those transformations are applied, such as converting data formats, filtering out specific information, or enriching the payload with additional context.

> 4. **Data buffering**: EMQX provides an on-disk message buffer to prevent data loss when the IoTDB is unavailable. Data is temporarily held in the buffer.

Reviewer: ```suggestion
4. **Data buffering**: EMQX provides an in-memory message buffer to prevent data loss when the IoTDB is unavailable. Data is temporarily held in the buffer, and may be offloaded to disk to prevent memory overload. Note that data is not preserved if the data integration or the EMQX node is restarted.
```

5. **Data ingestion into IoTDB**: Once the rule engine identifies a message for IoTDB storage, it triggers an action of forwarding the messages to IoTDB. Processed data will be seamlessly written into the IoTDB in a time series manner.

> 6. **Data Storage and Utilization**: With the data now stored in MongoDB, businesses can harness its querying power for various use cases. For instance, in the realm of connected vehicles, this stored data can inform fleet management systems about vehicle health, optimize route planning based on real-time metrics, or track assets. Similarly, in IIoT settings, the data might be used to monitor machinery health, forecast maintenance, or optimize production schedules.

Reviewer: ```suggestion
6. **Data Storage and Utilization**: With the data now stored in IoTDB, businesses can harness its querying power for various use cases. For instance, in the realm of connected vehicles, this stored data can inform fleet management systems about vehicle health, optimize route planning based on real-time metrics, or track assets. Similarly, in IIoT settings, the data might be used to monitor machinery health, forecast maintenance, or optimize production schedules.
```

     "root/#"
   ```

5. Click the **Add Action** button, and select **Forwarding with Data Bridge** from the dropdown list. With this action, EMQX sends the data processed by the rule to the data bridge.

6. Click the **+** icon next to the **Data bridge** drop-down box to create a data bridge.

7. Select **Apache IoTDB** from the **Type of Data Bridge** drop-down list. Fill in the required fields (marked with an asterisk).

8. Enter a name for the data bridge. The name should be a combination of upper/lower case letters and numbers.

9. Enter the connection information:

   * **IoTDB REST Service Base URL**: Enter `http://localhost:18080`, or the actual hostname/IP if the IoTDB server is running remotely.

   * **IoTDB Version**: Select the version of the IoTDB server you have installed.

>    * **Username**: Enter the IoTDB username; The default value is `root`.

Reviewer: ```suggestion
   * **Username**: Enter the IoTDB username.
```

>    * **Password**: Enter the IoTDB password; The default value is `root`.

Reviewer: ```suggestion
   * **Password**: Enter the IoTDB password.
```

   * **Device ID** (optional): Enter a fixed device id used as the name of the device from which the timeseries data is forwarded and inserted into the IoTDB instance.

     :::tip

>      When left empty, the device id can also be specified in the publishing message, configured in the rule, or extracted from the topic to which these messages are being sent by converting '/' into '.'. For example, publishing a message to `root/sg27` will result in sending a device name of `root.sg27`. However, the fixed device id configured in this field takes precedence over any methods mentioned previously.

Reviewer: @sstrigler may correct me, but I can't see this extraction from topic in the current code.

```suggestion
     When left empty, the device id can also be specified in the publishing message or configured in the rule. For example, when publishing a JSON-encoded message with a `device_id` field in it, the value of that field will define the output device id. To define it using the rule engine, you must use SQL like `select json_decode(payload) as payload, 'my_device' as payload.device_id`. However, the fixed device id configured in this field takes precedence over any methods mentioned previously.
```

     :::

>    - **Align Timeseries**: Disabled by default.

Reviewer: Found a reference that might be useful: https://iotdb.apache.org/UserGuide/V1.1.x/Data-Concept/Data-Model-and-Terminology.html#aligned-timeseries

Author: Can we write something like the following? Is that the correct purpose of this option?

**Align Timeseries**: Disabled by default. Once enabled, the timestamp columns of a group of aligned timeseries are stored only once in memory and on disk during data insertion, rather than duplicating them for each individual timeseries within the group. For more information, see [Aligned timeseries](https://iotdb.apache.org/UserGuide/V1.1.x/Data-Concept/Data-Model-and-Terminology.html#aligned-timeseries).

Reviewer: Well, this is a property of IoTDB rather than EMQX, so I wonder if we shouldn't just refer the user to its documentation.  The user probably knows more about its operations if they are using it.

Other than that, the phrase seems ok (I'm not familiar with IoTDB operations), but I'd just suggest a rephrasing:

```suggestion
   - **Align Timeseries**: Disabled by default. Once enabled, the timestamp columns of a group of aligned timeseries are stored only once in IoTDB, rather than duplicating them for each individual timeseries within the group. For more information, see [Aligned timeseries](https://iotdb.apache.org/UserGuide/V1.1.x/Data-Concept/Data-Model-and-Terminology.html#aligned-timeseries).
```

>    - **Enable TLS**: If you want to establish an encrypted connection, click the toggle switch. For more information about TLS connection, see [TLS for External Resource Access](../network/overview.md/#tls-for-external-resource-access).

Reviewer: ```suggestion
   - **Enable TLS**: If you want to establish an encrypted connection, click the toggle switch. For more information about TLS connection, see [TLS for External Resource Access](../network/overview.md#tls-for-external-resource-access).
```

| Query Mode               | Allows you to choose `asynchronous` or `synchronous` query modes to optimize message transmission based on different requirements. In asynchronous mode, writing to IoTDB does not block the MQTT message publish process. However, this might result in clients receiving messages ahead of their arrival in IoTDB. | `Async`            |
| Inflight Window          | An "in-flight query" refers to a query that has been initiated but has not yet received a response or acknowledgment. This setting controls the maximum number of in-flight queries that can exist simultaneously when the data bridge is communicating with Apache IoTDB. <br />When the `query_mode` is set to `async` (asynchronous), the "Inflight Window" parameter gains special importance. If it is crucial for messages from the same MQTT client to be processed in strict order, you should set this value to 1. | `100`              |

## More Information

> EMQX provides bunches of learning resources on the data integration with Apache Kafka. Check out the following links to learn more:

Reviewer: ```suggestion
EMQX provides bunches of learning resources on the data integration with Apache IoTDB. Check out the following links to learn more:
```
